pkg/receive/handler_test.go

```diff
@@ -94,47 +87,137 @@ func TestCountCause(t *testing.T) {
 				errors.New("foo"),
 				errors.New("bar"),
 			}),
-			f:   isConflict,
-			out: 1,
+			threshold: 1,
+			exp:       errConflict,
+		},
+		{
+			name: "matching but below threshold multierror",
+			err: terrors.MultiError([]error{
+				storage.ErrOutOfOrderSample,
+				errors.New("foo"),
+				errors.New("bar"),
+			}),
+			threshold: 2,
+			exp:       errors.New("unexpected error: 3 errors: out of order sample; foo; bar"),
 		},
 		{
 			name: "matching multierror many",
 			err: terrors.MultiError([]error{
 				storage.ErrOutOfOrderSample,
-				conflictErr,
-				errors.New(strconv.Itoa(http.StatusConflict)),
+				errConflict,
+				status.Error(codes.AlreadyExists, "conflict"),
 				errors.New("foo"),
 				errors.New("bar"),
 			}),
-			f:   isConflict,
-			out: 3,
+			threshold: 1,
+			exp:       errConflict,
+		},
+		{
+			name: "matching multierror many, one above threshold",
+			err: terrors.MultiError([]error{
+				storage.ErrOutOfOrderSample,
+				errConflict,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				errors.New("foo"),
+			}),
+			threshold: 2,
+			exp:       errNotReady,
+		},
+		{
+			name: "matching multierror many, both above threshold, conflict have precedence",
+			err: terrors.MultiError([]error{
+				storage.ErrOutOfOrderSample,
+				errConflict,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				status.Error(codes.AlreadyExists, "conflict"),
+				errors.New("foo"),
+			}),
+			threshold: 2,
+			exp:       errConflict,
 		},
 		{
 			name: "nested matching multierror",
-			err: errors.Wrap(terrors.MultiError([]error{
+			err: errors.Wrap(errors.Wrap(terrors.MultiError([]error{
 				storage.ErrOutOfOrderSample,
 				errors.New("foo"),
 				errors.New("bar"),
+			}), "baz"), "qux"),
+			threshold: 1,
+			exp:       errConflict,
+		},
+		{
+			name: "deep nested matching multierror",
```

Comment:
I'm just noting that this is new functionality compared to how error handling worked before